### PR TITLE
Generic Sort -- Remaining implementations

### PR DESCRIPTION
**Pull request by yoni206**
This PR completes the implementation of Generic Sorts. It also renames the generic solver's test to a generic sort test and adds new test cases for it.

### CODE REVIEW

File: src/generic_sort.cpp [hunk 1]
```diff
@@ -22,6 +22,104 @@ using namespace std;
 
 namespace smt {
 
+Sort make_uninterpreted_generic_sort(string name, uint64_t arity)
+{
+  return std::make_shared<UninterpretedGenericSort>(name, arity);
```
Comment (ahmed-irfan) [inline src/generic_sort.cpp:27]:
```suggestion
  return make_shared<UninterpretedGenericSort>(name, arity);
```

File: src/generic_sort.cpp [hunk 2]
```diff
@@ -22,6 +22,104 @@ using namespace std;
...
+}
+
+Sort make_uninterpreted_generic_sort(Sort sort_cons,
+                                     const SortVec & sorts) {
+  return std::make_shared<UninterpretedGenericSort>(sort_cons, sorts);
```
Comment (ahmed-irfan) [inline src/generic_sort.cpp:32]:
```suggestion
  return make_shared<UninterpretedGenericSort>(sort_cons, sorts);
```

File: src/generic_sort.cpp [hunk 3]
```diff
@@ -22,6 +22,104 @@ using namespace std;
...
+}
+
+
+Sort make_generic_sort(SortKind sk)
+{
+  if (sk != BOOL && sk != INT && sk != REAL)
+  {
+    throw IncorrectUsageException("Can't create sort from " + to_string(sk));
+  }
+  return std::make_shared<GenericSort>(sk);
```
Comment (ahmed-irfan) [inline src/generic_sort.cpp:42]:
```suggestion
  return make_shared<GenericSort>(sk);
```

File: src/generic_sort.cpp [hunk 4]
```diff
@@ -22,6 +22,104 @@ using namespace std;
...
+}
+
+Sort make_generic_sort(SortKind sk, uint64_t width)
+{
+  if (sk != BV)
+  {
+    throw IncorrectUsageException("Can't create sort from " + to_string(sk)
+                                  + " and " + ::std::to_string(width));
+  }
+  return std::make_shared<BVGenericSort>(width);
```
Comment (ahmed-irfan) [inline src/generic_sort.cpp:52]:
```suggestion
  return make_shared<BVGenericSort>(width);
```

File: src/generic_sort.cpp [hunk 5]
```diff
@@ -22,6 +22,104 @@ using namespace std;
...
+  throw IncorrectUsageException(
+      "No currently supported sort is created with a single sort argument");
+}
+
+Sort make_generic_sort(SortKind sk, Sort sort1, Sort sort2)
+{
+  Sort genericsort;
+  if (sk == ARRAY)
+  {
+    genericsort = std::make_shared<ArrayGenericSort>(sort1, sort2);
```
Comment (ahmed-irfan) [inline src/generic_sort.cpp:66]:
```suggestion
    genericsort = make_shared<ArrayGenericSort>(sort1, sort2);
```

File: src/generic_sort.cpp [hunk 6]
```diff
@@ -22,6 +22,104 @@ using namespace std;
...
+  }
+  else if (sk == FUNCTION)
+  {
+    genericsort =
+        std::make_shared<FunctionGenericSort>(SortVec{ sort1 }, sort2);
```
Comment (ahmed-irfan) [inline src/generic_sort.cpp:71]:
```suggestion
        make_shared<FunctionGenericSort>(SortVec{ sort1 }, sort2);
```

File: src/generic_sort.cpp [hunk 7]
```diff
@@ -22,6 +22,104 @@ using namespace std;
...
+                                  + sort2->to_string());
+  }
+  return genericsort;
+}
+
+Sort make_generic_sort(SortKind sk, Sort sort1, Sort sort2, Sort sort3)
+{
+  if (sk == FUNCTION)
+  {
+    return std::make_shared<FunctionGenericSort>(
```
Comment (ahmed-irfan) [inline src/generic_sort.cpp:86]:
```suggestion
    return make_shared<FunctionGenericSort>(
```

File: src/generic_sort.cpp [hunk 8]
```diff
@@ -22,6 +22,104 @@ using namespace std;
...
+  }
+}
+
+Sort make_generic_sort(SortKind sk, SortVec sorts)
+{
+  if (sk == FUNCTION)
+  {
+    Sort return_sort = sorts.back();
+    sorts.pop_back();
+    return std::make_shared<FunctionGenericSort>(sorts, return_sort);
```
Comment (ahmed-irfan) [inline src/generic_sort.cpp:103]:
```suggestion
    return make_shared<FunctionGenericSort>(sorts, return_sort);
```

File: src/generic_sort.cpp [hunk 9]
```diff
@@ -22,6 +22,104 @@ using namespace std;
...
+  }
+  else if (sk == ARRAY && sorts.size() == 2)
+  {
+    return std::make_shared<ArrayGenericSort>(sorts[0], sorts[1]);
```
Comment (ahmed-irfan) [inline src/generic_sort.cpp:107]:
```suggestion
    return make_shared<ArrayGenericSort>(sorts[0], sorts[1]);
```

File: src/generic_sort.cpp [hunk 10]
```diff
@@ -22,6 +22,104 @@ using namespace std;
...
+  }
+  else
+  {
+    std::string msg("Can't make sort from ");
```
Comment (ahmed-irfan) [inline src/generic_sort.cpp:111]:
```suggestion
    string msg("Can't make sort from ");
```

**Review by ahmed-irfan — COMMENTED** [review]
minor

**Review by makaimann — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**